lib/fdo.c

```diff
@@ -0,0 +1,1396 @@
+}
+
+/**
+ * Get pointer to memory holding device credentials which includes owner
+ * credentials and manufacturer credentials.
+ *
+ * @return ret
+ *        return pointer to memory holding device credentials on success, NULL
+ * if memory not allocated yet.
+ */
+fdo_dev_cred_t *app_get_credentials(void)
+{
+	return g_fdo_data->devcred;
+}
+
+/**
+ * Internal API
+ */
+static fdo_sdk_status app_initialize(void)
+{
+	int ret = FDO_ERROR;
+	int32_t fsize;
+	int max_serviceinfo_sz;
+	char *buffer = NULL;
+
+	if (!g_fdo_data)
+		return FDO_ERROR;
+
+	/* Initialize service_info to NULL in case of early error. */
+	g_fdo_data->service_info = NULL;
+
+/* Enable/Disable Error Recovery */
+#ifdef RETRY_FALSE
+	g_fdo_data->error_recovery = false;
+#else
+	g_fdo_data->error_recovery = true;
+#endif
+	g_fdo_data->recovery_enabled = false;
+	g_fdo_data->state_fn = &_STATE_TO1;
+	if (memset_s(&g_fdo_data->prot, sizeof(fdo_prot_t), 0) != 0) {
+		LOG(LOG_ERROR, "Memset Failed\n");
+		return FDO_ERROR;
+	}
+
+	g_fdo_data->err = 0;
+
+#ifdef CLI
+	/* Process command line input. */
+	ret = input_parameters(g_argc, g_argv);
+	if (0 != ret) {
+		return FDO_ERROR;
+	}
+#endif
+
+	// read the file at path MAX_SERVICEINFO_SZ_FILE to get the maximum ServiceInfo size
+	// that will be supported for both Owner and Device ServiceInfo
+	// default to MIN_SERVICEINFO_SZ if the file is empty/non-existent
+	// file of size 1 is also considered an empty file containing new-line character
+	fsize = fdo_blob_size((char *)MAX_SERVICEINFO_SZ_FILE, FDO_SDK_RAW_DATA);
+	if (fsize == 0 || fsize == 1) {
+		g_fdo_data->prot.maxDeviceServiceInfoSz = MIN_SERVICEINFO_SZ;
+		g_fdo_data->prot.maxOwnerServiceInfoSz = MIN_SERVICEINFO_SZ;
+		g_fdo_data->prot.prot_buff_sz = MIN_SERVICEINFO_SZ + MSG_METADATA_SIZE;
+	} else if (fsize > 0) {
+		buffer = fdo_alloc(fsize + 1);
+		if (buffer == NULL) {
+			LOG(LOG_ERROR, "malloc failed\n");
+		} else {
+			if (fdo_blob_read((char *)MAX_SERVICEINFO_SZ_FILE, FDO_SDK_RAW_DATA,
+					(uint8_t *)buffer, fsize) == -1) {
+				LOG(LOG_ERROR, "Failed to read Manufacture DN\n");
+			}
+			max_serviceinfo_sz = atoi(buffer);
+			if (max_serviceinfo_sz <= MIN_SERVICEINFO_SZ) {
+				max_serviceinfo_sz = MIN_SERVICEINFO_SZ;
+			}
+			else if (max_serviceinfo_sz >= MAX_SERVICEINFO_SZ) {
+				max_serviceinfo_sz = MAX_SERVICEINFO_SZ;
+			}
+			g_fdo_data->prot.prot_buff_sz = max_serviceinfo_sz + MSG_METADATA_SIZE;
+			g_fdo_data->prot.maxDeviceServiceInfoSz = max_serviceinfo_sz;
+			g_fdo_data->prot.maxOwnerServiceInfoSz = max_serviceinfo_sz;
+		}
+	}
+	if (buffer != NULL) {
+		fdo_free(buffer);
+	}
+
+	/* 
+	* Initialize and allocate memory for the FDOW/FDOR blocks before starting the spec's 
+	* protocol execution. Reuse the allocated memory by emptying the contents.
+	*/
+	if (!fdow_init(&g_fdo_data->prot.fdow) ||
+		!fdo_block_alloc_with_size(&g_fdo_data->prot.fdow.b,
+			g_fdo_data->prot.prot_buff_sz)) {
+		LOG(LOG_ERROR, "fdow_init() failed!\n");
```

Comment:
There is the possibility of a memory leak if fdow_init or fdow_init is failed then previously allocated buffer or memory is not deallocated